drivers/overlay/overlay.go

```diff
@@ -1486,3 +1595,80 @@ func nameWithSuffix(name string, number int) string {
 	}
 	return fmt.Sprintf("%s%d", name, number)
 }
+
+func (d *Driver) getAdditionalLayerPath(dgst digest.Digest, ref string) (string, error) {
+	refelem := base64.StdEncoding.EncodeToString([]byte(ref))
+	for _, ls := range d.options.layerStores {
+		var elems []string
+		if ls.withReference {
+			elems = append(elems, refelem)
+		}
+		elems = append(elems, dgst.String())
+		target := filepath.Join(append([]string{ls.path}, elems...)...)
+		// Check if all necessary files exist
+		for _, p := range []string{
+			filepath.Join(target, "diff"),
+			filepath.Join(target, "info"),
+			// TODO(ktock): We should have an API to expose the stream data of this layer
+			//              to enable the client to retrieve the entier contents of this
+			//              layer when it exports this layer.
+		} {
+			if _, err := os.Stat(p); err != nil {
+				return "", graphdriver.ErrLayerUnknown
+			}
+		}
+		return target, nil
+	}
+
+	return "", graphdriver.ErrLayerUnknown
+}
+
+type additionalLayer struct {
+	path      string
+	d         *Driver
+	closeOnce sync.Once
+}
+
+// Info returns arbitrary information stored along with this layer (i.e. `info` file).
+func (al *additionalLayer) Info() (io.ReadCloser, error) {
+	return os.Open(filepath.Join(al.path, "info"))
+}
+
+// CreateAs creates a new layer from this additional layer.
+func (al *additionalLayer) CreateAs(id, parent string) error {
+	targetDiffDir := filepath.Join(al.path, "diff")
+	if _, err := os.Stat(targetDiffDir); err != nil {
+		return err
+	}
+	if ld, err := os.Readlink(targetDiffDir); err == nil {
+		// `diff` can be symlink
+		targetDiffDir = ld
+	}
+
+	// TODO: support opts
+	if err := al.d.Create(id, parent, nil); err != nil {
+		return err
+	}
+	dir := al.d.dir(id)
+	// mark this layer as "additional layer"
+	if err := ioutil.WriteFile(path.Join(dir, "additionallayer"), []byte(al.path), 0644); err != nil {
+		return err
+	}
+	diffDir := path.Join(dir, "diff")
+	if err := os.RemoveAll(diffDir); err != nil {
+		return err
+	}
+	// tell the additional layer store that we use this layer.
+	if f, err := os.Create(path.Join(al.path, "use")); err == nil {
+		f.Close()
+	}
+	return os.Symlink(targetDiffDir, diffDir)
+}
+
+// Release tells the additional layer store that we don't use this handler.
+func (al *additionalLayer) Release() {
+	al.closeOnce.Do(func() {
```

Comment:
1. Why is this called `closeOnce` it the `Rmdir` (but not `Close`) is wrapped into it?

2. I don't understand why the syncOnce here? Is repeated `Rmdir` can cause some harm?

Reply:
Please see also: https://github.com/containers/storage/pull/795#issuecomment-771594370

c/storage sends notification when it starts to use a layer and when it releases that layer. Additional layer store can internally manage reference counter based on these notifications and can perform GC when nobody refers to a layer. Creating `<Store>/<layerdigest>/use` file notifies that this layer is used and rmdir-ing `<Store>/<layerdigest>` notifies that this layer is no longer used.

The mentioned `Rmdir` decreases the reference counter originally increased by *`(c/storage/drivers/overlay).Driver.LookupAdditionalLayer()`. Calling this twice sends notification twice, which breaks reference management on the Additional Layer Store's side.